src/crypto/openssl/cose_sign.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+
+#include "crypto/openssl/cose_sign.h"
+
+#include "ccf/ds/logger.h"
+
+#include <openssl/evp.h>
+#include <t_cose/t_cose_sign1_sign.h>
+
+namespace
+{
+  constexpr int64_t COSE_HEADER_PARAM_ALG =
+    1; // Duplicate of t_cose::COSE_HEADER_PARAM_ALG to keep it compatible.
+
+  size_t estimate_buffer_size(
+    const std::unordered_map<int64_t, std::string>& protected_headers,
+    std::span<const uint8_t> payload)
+  {
+    size_t result =
+      300; // bytes for metadata even everything else is empty. This's the most
+           // often used value in the t_cose examples, however no recommendation
+           // is provided which one to use. We will consider this an affordable
+           // starting point, as soon as we don't expect a shortage of memory on
+           // the target platforms.
+
+    result = std::accumulate(
+      protected_headers.begin(),
+      protected_headers.end(),
+      result,
+      [](auto result, const auto& kv) {
+        return result + sizeof(kv.first) + kv.second.size();
+      });
+
+    return result + payload.size();
+  }
+
+  void encode_protected_headers(
+    t_cose_sign1_sign_ctx* ctx,
+    QCBOREncodeContext* encode_ctx,
+    const std::unordered_map<int64_t, std::string>& protected_headers)
+  {
+    QCBOREncode_BstrWrap(encode_ctx);
+    QCBOREncode_OpenMap(encode_ctx);
+
+    // This's what the original implementation of `encode_protected_parameters`
```

Comment:
```suggestion
    // This's what the t_cose implementation of `encode_protected_parameters`
```

Please add a comment on each function that's a fork of a t_cose version stating that it is and how it differs from the original.